cloudevents/working-drafts/dds-format.md

```diff
@@ -0,0 +1,247 @@
+messages transmitted over the DDS protocol are defined via the IDL type system.
+These messages are sent over the wire using the Common Data Representation (CDR)
+serialization, as defined in the OMG [Real-Time Publish Subscribe (RTPS)][rtps]
+specification.
+
+## Table of Contents
+
+1. [Introduction](#1-introduction)
+2. [Attributes](#2-attributes)
+3. [Data](#3-data)
+4. [Transport](#4-transport)
+5. [Batch Format](#5-batch-format)
+6. [Examples](#6-examples)
+
+## 1. Introduction
+
+[CloudEvents][ce] is a standardized and protocol-agnostic definition of the
+structure and metadata description of events. This specification defines how the
+elements defined in the CloudEvents specification are are represented using
+a protobuf schema.
+
+The [Attributes](#2-attributes) section describes the naming conventions and
+data type mappings for CloudEvent attributes for use as protobuf message
+properties.
+
+The [Data](#3-data) section describes how the event payload is carried.
+
+### 1.1. Conformance
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][rfc2119].
+
+### 1.2 Content-Type
+
+There is no official IANA *media-type* designation for DDS, as such this
+specification uses 'application/dds' to identify such content.
+
+## 2. Attributes
+
+This section defines how CloudEvents attributes are represented in the DDS
+[schema][dds-schema].
+
+## 2.1 Type System
+
+The CloudEvents type system is mapped to IDL datatypes as follows :
+
+| CloudEvents   | DDS |
+| ------------- | ---------------------------------------------------------------------- |
+| Boolean       | [boolean][dds-schema] |
+| Integer       | [int32][dds-schema] |
+| String        | [string][dds-schema] |
+| Binary        | [bytes][dds-schema] |
+| URI           | [string][dds-schema] following [RFC 3986 §4.3][rfc3986-section43]|
+| URI-reference | [string][dds-schema] following [RFC 3986 §4.1][rfc3986-section41] |
+| Timestamp     | [Timestamp][dds-schema]  |
+
+## 2.3 REQUIRED Attributes
+
+REQUIRED attributes are represented explicitly as fields in the [Event type definition][dds-schema].
+
+## 2.4 OPTIONAL Attributes & Extensions
+
+OPTIONAL and extension attributes are represented using a sequence of key-value
+constructs (DDS-CE-Attributes) enabling direct support of the CloudEvent
+[type system][ce-types]. An underlying union and enumeration data structure
+are needed to construct the key-value sequence to establish the association
+between the IDL types and the CloudEvent type system.
+
+```xml
+
+  <enum name="AttributeType">
+   <enumerator name="BOOL"/>
+   <enumerator name="INT32"/>
+   <enumerator name="STRING"/>
+   <enumerator name="BYTES"/>
+   <enumerator name="URI"/>
+   <enumerator name="URI_REF"/>
+   <enumerator name="TIMESTAMP"/>
+ </enum>
+	  
+ <union name="AttributeValue" extensibility="final">
```

Comment:
Do you need to add a <case> to represent a Timestamp type ???

Reply:
Same as above. (But this was actually a copy/paste error from my sample code. I forgot the timestamp for some reason.)